enhancements/network/primary-user-defined-networks-for-virtualization-workloads.md

```diff
@@ -0,0 +1,638 @@
+
+### Persisting VM IP addresses during the migration
+
+OpenShift already features the ability of providing persistent IP addresses
+for layer2 **secondary** user defined networks. It relies on the
+[IPAMClaim CRD](https://github.com/k8snetworkplumbingwg/ipamclaims/blob/48c5a915da3b67f464a4e52fa50dbb3ef3547dcd/pkg/crd/ipamclaims/v1alpha1/types.go#L23)
+to "tie" the allocation of the IP address to the VM.
+In short, OpenShift Virtualization creates an `IPAMClaim` resource for each
+interface in the VM the user wants to be persistent, and then instructs
+the CNI plugin of which claim to use to persist the IP for the VM using an
+attribute defined in the k8snetworkplumbingwg network-selection-elements (the
+`k8s.v1.cni.cncf.io/networks` annotation). This protocol is described in depth
+in
+[the persistent IPs for virt workloads enhancement](https://github.com/openshift/enhancements/blob/master/enhancements/network/persistent-ips-sdn-secondary-networks.md#creating-a-virtual-machine).
+
+We want to re-use this mechanism to implement persistent IPs for UDN networks;
+meaning OpenShift Virtualization will create the `IPAMClaim` and will instruct
+OVN-Kubernetes of which claim to use. Since for primary UDNs we do **not**
+define the network-selection-elements, we need to use a new annotation to pass
+this information along.
+
+The proposed annotation value is `k8s.ovn.org/primary-udn-ipamclaim`.
+
+These persistent IPs will be cleaned up by the Kubernetes garbage collector
+once the VM to which they belong is removed.
+
+All the other
+[work required in OVN-Kubernetes](https://github.com/trozet/enhancements/blob/941f5c6391830d5e4a94e65d742acbcaf9b8eda9/enhancements/network/user-defined-network-segmentation.md#pod-egress)
+to support live-migration was already implemented as part of the epic
+implementing UDN.
+
+We evaluate an alternative for this in the
+[Using multus default network annotation](#using-multus-default-network-annotation)
+section.
+
+### Allow user to configure the VMs interface desired IP address
+
+To allow the user to specify the IP address for the workload we plan on adding
+an attribute to the KubeVirt API
+[interface](https://kubevirt.io/api-reference/main/definitions.html#_v1_interface)
+named `ipAddress`. This approach suits the product, and meets the user's
+expectations because the user can configure the MAC address for an interface in
+that same `Interface` resource.
+
+Regarding the implementation, OpenShift Virtualization would extend what it
+does for the MAC address: define it in the k8snetworkplumbingwg de-facto
+standard as an IP request, and the OVN-Kubernetes CNI would attempt to honor
+it.
+
+This is by far the simplest solution to implement; another alternative is
+described in the [alternatives section](#vm-interface-ip-address-configuration).
+This other alternative has the advantage of not requiring a KubeVirt API update
+but adds a lot more work in OVN-Kubernetes, and requires updating the de-facto
+standard.
+
```

Comment:
Another option:

```suggestion
Another option is to manage IP assignments on `UserDefinedNetwork`, where IPs from the excluded subnet could be assigned to a VM or a Pod by matching it's namespace and name.

The benefits of this are most obvious when looking at `ClusterUserDefinedNetwork`. Keeping all the mappings at a single place would make it easier to get a full picture of free and assigned IPs. Keeping the mapping on the UDN may be also considered more secure - the network belongs to the person defining the subnet, in case of `CUDN` to the cluster-admin. With this approach it will be always this owner giving away static IPs. This mimicks the workflow in traditional networks (reserving IPs on DHCP server configuration), and also the API used by VMware.

The implementation could look like this:

* We keep IPAMClaims as they are
* UDN now has a new section where I can assign IPs from excluded subnet to Pods/VMs
* When Pod is created, OVN checks if it has reserved IP on UDN, if it does it uses it
```

Reply:
This is interesting; it approaches the problem in a different way: I had assumed we would want the VM owner to be able to specify the IP; this approach moves that responsibility / ability to the admin.

Without getting into the technical limitations we would face for it (now) it is quite clear we need to clarify which persona we want to be able to indicate the VM's IPs - the admin, or the VM owner.

I've reached some stake holders / field people. I'll clarify once I have heard from them.

Until we know exactly who the "user" is for that, no point